docs/src/main/asciidoc/extension-writing-dev-service.adoc

```diff
@@ -0,0 +1,123 @@
+                                         .withReuse(true);
+
+        container.start();
+
+        String newUrl = "http://" + container.getHost() + ":" + container.getMappedPort(SERVICE_PORT);
+        Map<String, String> configOverrides = Map.of("some-service.base-url", newUrl);
+
+        return new DevServicesResultBuildItem.RunningDevService(FEATURE, container.getContainerId(),
+            container::close, configOverrides)
+            .toBuildItem();
+    }
+----
+
+With this code, you should be able to see your container starting if you add your extension to a test application and run `quarkus dev`.
+However, the application will not be able to connect to it, because no ports are exposed. To expose ports, add `withExposedPorts` to the container construction.
+For example,
+
+[source%nowrap,java]
+----
+GenericContainer container = new GenericContainer<>(dockerImageName)
+                                 .withExposedPorts(SERVICE_PORT, OTHER_SERVICE_PORT);
+----
+
+Testcontainers will map these ports to random ports on the host. This avoids port conflicts, but presents a new problem – how do applications connect to the service in the container?
+
+To allow applications to connect, the extension should override the default configuration for the service with the mapped ports.
+This must be done after starting the container.
+For example,
+
+[source%nowrap,java]
+----
+    container.start();
+    Map<String, String> configOverrides = Map.of("some-service.base-url",
+         "http://" + container.getHost() + ":" + container.getMappedPort(SERVICE_PORT));
+----
+
+Other configuration overrides may be included in the same map.
+
+== Waiting for the container to start
+
+You should add a `.waitingFor` call to the container construction, to wait for the container to start. For example
+
+[source%nowrap,java]
+----
+            .waitingFor(Wait.forLogMessage(".*" + "Started" + ".*", 1))
+----
+
+Waiting for a port to be open is another option. See the link:https://java.testcontainers.org/features/startup_and_waits/[Testcontainers documentation] for a full discussion of wait strategies.
+
+== Configuring the dev service
+
+To configure the dev service launch process, your build step can accept a `ConfigPhase.BUILD_TIME` config class in its constructor.
+For example,
+
+[source%nowrap,java]
+----
+    @BuildStep(onlyIfNot = IsNormal.class, onlyIf = GlobalDevServicesConfig.Enabled.class) {
+    public DevServicesResultBuildItem createContainer(MyConfig config) {
+----
+
+You may wish to use this config to set a fixed port, or set an image name, for example.
+
+[source%nowrap,java]
+----
+    if (config.port.isPresent()) {
+        container.setPortBindings(List.of(config.port.get() + ":" + SERVICE_PORT));
+    }
+----
+
+== Controlling re-use
+
+In dev mode, with hot reload, Quarkus may restart frequently. By default, this will also restart test containers.
```

Comment:
```suggestion
In dev mode, with live reload, Quarkus may restart frequently. By default, this will also restart test containers.
```

Reply:
Oh, thank you! I had taken hot reload from one of our sources (code comments or docs or something - apparently we have 43 references to it in our codebase), and then I realised I should go back and fix it. But, apparently, did not.